lib/semantics/tools.cc

```diff
@@ -382,4 +354,26 @@ const Symbol *HasEventOrLockPotentialComponent(
   }
   return nullptr;
 }
+
+const Symbol *FindUltimateComponent(const DerivedTypeSpec &derivedTypeSpec,
+    std::function<bool(const Symbol &)> predicate) {
+  const auto *scope{derivedTypeSpec.typeSymbol().scope()};
+  CHECK(scope);
+  for (const auto &pair : *scope) {
+    const Symbol &component{*pair.second};
+    const DeclTypeSpec *type{component.GetType()};
+    if (!type) {
+      continue;
+    }
+    const DerivedTypeSpec *derived{type->AsDerived()};
+    bool isUltimate{IsAllocatableOrPointer(component) || !derived};
+    if (const Symbol *
+        result{!isUltimate ? FindUltimateComponent(*derived, predicate)
```

Comment:
Consider reversing the predicate to avoid the negation, for readability.

Reply:
It is currently in the form `!a ? x : b ? y : z`.
Inverting the first condition makes it `a ? b ? y : z : x`.
I think the latter is less readable.